src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.RegularCompilationTracker_Generators.cs

```diff
@@ -36,49 +36,71 @@ private sealed partial class RegularCompilationTracker : ICompilationTracker
             Compilation? compilationWithStaleGeneratedTrees,
             CancellationToken cancellationToken)
         {
-            if (creationPolicy.GeneratedDocumentCreationPolicy is GeneratedDocumentCreationPolicy.DoNotCreate)
+            // First try to compute the SG docs in the remote process (if we're the host process), syncing the results
+            // back over to us to ensure that both processes are in total agreement about the SG docs and their
+            // contents.
+            var result = await TryComputeNewGeneratorInfoInRemoteProcessAsync(
+                compilationState, compilationWithoutGeneratedFiles, generatorInfo.Documents, compilationWithStaleGeneratedTrees, creationPolicy.GeneratedDocumentCreationPolicy, cancellationToken).ConfigureAwait(false);
+            if (result.HasValue)
             {
-                // We're frozen.  So we do not want to go through the expensive cost of running generators.  Instead, we
-                // just whatever prior generated docs we have.
-                var generatedSyntaxTrees = await generatorInfo.Documents.States.Values.SelectAsArrayAsync(
-                    static (state, cancellationToken) => state.GetSyntaxTreeAsync(cancellationToken), cancellationToken).ConfigureAwait(false);
+                var updatedCompilationWithGeneratedFiles = result.Value.compilationWithGeneratedFiles;
+                var updatedGeneratedDocuments = result.Value.generatedDocuments;
 
-                var compilationWithGeneratedFiles = compilationWithoutGeneratedFiles.AddSyntaxTrees(generatedSyntaxTrees);
```

Comment:
This seems like a bug that we weren't returning compilationWithStaleGeneratedTrees if it existed in this case. This would have meant we were potentially throwing out decltrees for the generated files when we didn't need to be. Obviously you're deleting the code but do think through if we're making this better or worse.